feat: support passive events by define global variables in zone.js config file

JiaLiPassion opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #8866

It is not easy to `addEventListener` with `passive` option. We may need to discuss to add something like in this PR, https://github.com/angular/angular/pull/21681 in the future.

## What is the new behavior?
For now, this PR is just a temp solution until we can add a new `template syntax` in `event listener`.
In this PR, we can define a global variable before importing zone.js

```
(window as any)['__zone_symbol__PASSIVE_EVENTS'] = ['touchstart', 'scroll'];
```

and then all event listeners defined here will be automatically be `passive: true`.
## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

If this idea is ok, I will update the document. And of course we still need to discuss the formal way to resolve this issue by introducing some new options in `template event binding` syntax.

JiaLiPassion commented:
@mhevery, please check whether this temp solution is ok, thank you!

mhevery commented on packages/zone.js/test/browser/browser.spec.ts:
```diff
@@ -1035,6 +1037,42 @@ describe('Zone', function() {
+      describe('passiveEvents', () => {
+        let logs: string[] = [];
+        const listener = (e: Event) => {
+          logs.push(e.defaultPrevented.toString());
```
Instead of pushing `true`/`false` which is hard to understand in the assertion, could we push text
```suggestion
          logs.push(e.defaultPrevented ? 'defaultPrevented' : 'default will run');
```

JiaLiPassion replied:
Yeah, this is better, updated.

mhevery commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -363,12 +383,26 @@ export function patchEventTarget(
         const options = arguments[2];
+        let passive = false;
+        if (passiveSupported && !!passiveEvents) {
+          for (let i = 0; i < passiveEvents.length; i++) {
```
why not just use `passiveEvents.indexOf(...)`?

JiaLiPassion replied:
Yeah, you are right, updated. Thanks.

mhevery commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -247,6 +247,25 @@ export function patchEventTarget(
       }
     }
 
+    function buildPassiveOptions(options: any, passive: boolean) {
```
Could we add some documentation on this method. What is its purpose? why do we need it?

JiaLiPassion replied:
Sure, this method is to build an `option object` to handle all possible inputs. doc added, thanks.

mhevery commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -363,12 +383,26 @@ export function patchEventTarget(
           for (let i = 0; i < blackListedEvents.length; i++) {
             if (eventName === blackListedEvents[i]) {
-              return nativeListener.apply(this, arguments);
+              if (!passive) {
```
can you flip `!passive` to `passive` and just flip the `else` branch?

JiaLiPassion replied:
Yeah, that will be natural. Thanks.

mhevery commented:
Looks like a reasonable approach.

JiaLiPassion commented:
@mhevery, thanks for the review, I have updated the PR and add the document.

mhevery commented on packages/zone.js/test/browser/browser.spec.ts:
```diff
@@ -1035,6 +1037,43 @@ describe('Zone', function() {
+        };
+        beforeEach(() => { logs = []; });
+        it('should be passive with global variable defined',
+           () => { testPassive('touchstart', true, {passive: true}); });
```
also change `true`/`false` in these tests methods to the actual string. It will read much better.

JiaLiPassion replied:
Got it, thanks.

mhevery commented on integration/_payload-limits.json:
```diff
@@ -3,8 +3,8 @@
         "runtime-es2015": 1485,
-        "main-es2015": 141000,
-        "polyfills-es2015": 36657
+        "main-es2015": 141476,
```
Why are these getting bigger?

JiaLiPassion replied:
I have checked the diff between the old and new dist bundles, the differences are exactly the source added to handle passive in this PR, such as this `buildPassiveOptions` function.

mhevery commented on packages/zone.js/test/browser/browser.spec.ts:
```diff
@@ -222,6 +222,7 @@ describe('Zone', function() {
             });
 
             zone.run(() => { document.dispatchEvent(scrollEvent); });
+            (document as any).removeAllListeners('scroll');
```
Why do we have to cast this to `any`?

JiaLiPassion replied:
Because `removeAllListeners` are the API added by `zone.js`, it is not a standard API.

mhevery replied:
Could we create an interface such as 
```
interface ZoneDocumentExtensions {
  /** Docs here */
  removeAllListeners(name: string): void;
}
```

That way it would be clear which methods we are adding and what their docs are? We could than publish these interfaces.

mhevery replied:
(I am OK if you want to do this in followup PR)

JiaLiPassion replied:
Yeah, sure, this is a good idea, there are several other APIs zone.js added, I will create another PR to add it, thanks.

JiaLiPassion commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -247,6 +247,27 @@ export function patchEventTarget(
       }
     }
 
+    // this util function will build an option object with passive option
```
Got it, updated, thanks!

mhevery commented:
[presubmit](http://test/OCL:294461424:BASE:294461465:1581443932965:253ce40f)

JiaLiPassion commented:
@kara, I have rebased the PR, please review, thanks.

mhevery commented:
[presubmit](http://test/OCL:294461424:BASE:296256105:1582226487454:ca5d889e)

kara commented on aio/content/guide/user-input.md:
```diff
@@ -297,7 +297,24 @@ Following is all the code discussed in this page.
 
+### Passive events
 
+Angular also support passive event listeners. For example, we can use the following steps to make scroll event passive.
```
support => supports

```
Angular also supports passive event listeners. 
```

JiaLiPassion replied:
Got it, updated, thanks.

kara commented on aio/content/guide/user-input.md:
```diff
@@ -297,7 +297,24 @@ Following is all the code discussed in this page.
 
+Angular also support passive event listeners. For example, we can use the following steps to make scroll event passive.
+
+1. create a file `src/zone-flags.ts`.
```
create -> Create

Can we capitalize all first words of sentences in this new section?

JiaLiPassion replied:
Got it, updated, thanks.

kara commented on aio/content/guide/user-input.md:
````diff
@@ -297,7 +297,24 @@ Following is all the code discussed in this page.
+```
+(window as any)['__zone_symbol__PASSIVE_EVENTS'] = ['scroll'];
+```
+3. in `src/polyfills.ts`, before import zone.js, import the new created `zone-flags`.
````
before import zone.js => before importing zone.js

JiaLiPassion replied:
Got it, updated, thanks.

kara commented on aio/content/guide/user-input.md:
````diff
@@ -297,7 +297,24 @@ Following is all the code discussed in this page.
+import 'zone.js/dist/zone';  // Included with Angular CLI.
+```
+
+after those steps, if user add event listeners for the `passive` event, the listeners will be `passive`.
````
if user add event listeners => if user adds event listeners

kara commented on integration/_payload-limits.json:
```diff
@@ -4,7 +4,7 @@
         "runtime-es2015": 1485,
         "main-es2015": 141569,
-        "polyfills-es2015": 36657
+        "polyfills-es2015": 37300
```
This is a somewhat large increase. Do we know why? Can we find out how much is just accumulation?

IgorMinar replied:
I agree, this seems excessive especially for the diff in this PR. Let's find out how did we accumulate this much (I've reset the polyfills limits recently so it seems like a recent accumulation of code)

kara reviewed (commented):
Can we also add a descriptive commit message that explains why this change is necessary?

mhevery commented:
This was accidentally merged. Reverting until we have better review.

JiaLiPassion committed:
feat: add a temp solution to support passive event listeners.

Now Angular doesn't support add event listeners as passive very easily.
User needs to use `elem.addEventListener('scroll', listener, {passive: true});`
or implements their own EventManagerPlugin to do that.
Angular may finally support new template syntax to support passive event, for now,
this commit introduces a temp solution to allow user to define the passive event names
in zone.js configurations.

User can define a global varibale like this.

```
(window as any)['__zone_symbol__PASSIVE_EVENTS'] = ['scroll'];
```

to let all `scroll` event listeners passive.

JiaLiPassion commented:
@kara, @mhevery , I have updated the doc and remove some unused code. Now the size will have no changes. Please review, thanks.

mhevery commented:
[presubmit](http://test/OCL:294461424:BASE:296967878:1582581629409:8e2a6409)

kara commented on aio/content/guide/user-input.md:
```diff
@@ -298,7 +298,23 @@ Following is all the code discussed in this page.
 </code-tabs>
 
 
+Angular also supports passive event listeners. For example, you can use the following steps to make scroll event passive.
```
```suggestion
Angular also supports passive event listeners. For example, you can use the following steps to make the scroll event passive.
```

JiaLiPassion replied:
Got it, thanks, updated.

kara commented on aio/content/guide/user-input.md:
````diff
@@ -298,7 +298,23 @@ Following is all the code discussed in this page.
+(window as any)['__zone_symbol__PASSIVE_EVENTS'] = ['scroll'];
+```
+
+3. In the `src/polyfills.ts` file, before importing zone.js, import the new created `zone-flags`.
````
```suggestion
3. In the `src/polyfills.ts` file, before importing zone.js, import the newly created `zone-flags`.
```

JiaLiPassion replied:
Got it, updated, thanks.

kara reviewed (changes requested):
Thanks for updating the PR! There are still a few things to do:

1) Update the commit message to describe what problem this PR is fixing and why (should not just be one sentence).

2) A few more docs updates below to fix

JiaLiPassion commented:
@kara, thank you for the review, I have updated the doc and the commit message. Please review.

kara reviewed (approved):
LGTM, thanks for updating the PR @JiaLiPassion!

sod commented:
What's the release schedule for zone.js? I ask as I saw quite some changes over time, but the last release was 8 months ago (0.10.2 from August 14, 2019 - https://yarnpkg.com/package/zone.js)

JiaLiPassion commented:
@sod, there should be a new version very soon.
